grails-app/controllers/streama/ReportController.groovy

```diff
@@ -12,18 +12,32 @@ import grails.transaction.Transactional
 class ReportController {
 
   static responseFormats = ['json', 'xml']
-  static allowedMethods = [save: "PUT"]
+//  static allowedMethods = [save: "PUT"]
 
   def springSecurityService
 
   def index () {
-    def jsonData = request.JSON
-
-    if (jsonData) {
-
+    def filter = params.filter
+    def responseObj = [
+      reports: [],
+      count: 0
+    ]
+
+    if(filter == 'all'){
+      responseObj.reports = Report.list(params)
+      responseObj.count = Report.count()
+    }
+    if(filter == 'unresolved'){
+      responseObj.reports = Report.list().findAll{!it.resolved}
+      responseObj.count = responseObj.reports?.size()
+    }
+    if(filter == 'resolved'){
+      responseObj.reports = Report.list().findAll{it.resolved}
+      responseObj.count = responseObj.reports?.size()
+    }
```

Comment:
there is a lot of redundant code here. Here is a different approach: 

```groovy
Boolean resolved = params.boolean('resolved')

List<Report> resports = Report.where{
if(resolved != null){
 resolved == resolved
}
}.list()

```

this way, you 
- only have 1 where query for all 3 cases
- you have a 1-to-1-matching for the param-name and the property to filter on (null equals "all" since the query is then not included, true equals "resolved", false equals "unresolved")
- this where query is easily extendable with more filters & sorting
- it is higher in performance, because your code basically says: get all reports from database, and THEN go through all and check the criteria inside of the `findAll`. my suggestion only fetches the relevant records from the db.

Reply:
Hmm, I looked at this a bit now and I think I need some explanation here. Not sure what you're doing here exactly.

Reply:
All right! will explain in a workshop

Reply:
```groovy
def index () {
    Boolean resolved = params.boolean('resolved')
    Long createdById = params.long('createdById')
    String sort = params.sort
    String order = params.order
    Integer max = params.int('max', 30)
    Integer offset = params.int('offset', 0)

    def responseObj = [
      reports: [],
      count: 0
    ]

    DetachedCriteria reportQuery = Report.where {
      if (resolved != null) {
        resolved == resolved
      }
      if (createdById != null) {
        createdBy.id == createdById
      }
    }
    responseObj.reports = reportQuery.list(sort: sort, order: order, max: max, offset: offset)
    responseObj.count = reportQuery.count()

    JSON.use('adminReports'){
      render (responseObj as JSON)
    }
  }

```